### PR TITLE
Read write ephemeral objects for remote publication of cluster state

### LINKED ISSUE #13728: [Remote Cluster State] Upload ephemeral cluster state objects(cluster blocks, nodes, customs) to remote
### Is your feature request related to a problem? Please describe

Meta issue: https://github.com/opensearch-project/OpenSearch/issues/13683
Ephemeral objects like cluster blocks , discovery nodes and ClusterState.custom objects are not uploaded to remote store currently.

### Describe the solution you'd like

Upload the ephemeral objects to remote store

### Related component

Cluster Manager

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #13729: [Remote Cluster State] Download the cluster state objects asynchronously and in parallel
### Is your feature request related to a problem? Please describe

Meta issue: https://github.com/opensearch-project/OpenSearch/issues/13683
Currently the remote cluster state objects are downloaded sequentially.

### Describe the solution you'd like

Download the remote cluster state objects asynchronously so that the downloads can be parallelized.

### Related component

Cluster Manager

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #14109: [Remote Routing Table] Add read flow from remote store
### Is your feature request related to a problem? Please describe

Add read flow for routing table from remote store. Meta issue: #12995 

### Describe the solution you'd like

..

### Related component

ShardManagement:Resiliency

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shiv0408**
### Description
The changes include modification in `RemoteClusterStateService` to enable uploading the ephemeral objects required for publication of cluster state through remote store. Also, added a method to read the Cluster State Attributes such as IndexMetadata, GlobalMetadata attributes from remote store in parallel. Read will be crucial part for publication as each follower node will be downloading the updated state from remote. 

### Related Issues
Resolves #13728
Resolves #13729 
Resolves #14109 

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:209]: We can remove unnecessary logs from PR

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1251]: For my understanding, what does this flag `includeEphemeral` mean?

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1031]: Do we need a validation on if the remote routing table is enabled or not here?

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1280]: Check for `remoteRoutingTableService.isPresent` before invoking

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java]: Added.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java]: Done.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java]: Done.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1253]: @shiv0408  @soosinha  can take this question better.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1253]: For publication of cluster state, we need ephemeral objects such as DiscoveryNodes, ClusterBlocks. Which are not required for just recovering the cluster state. This flag is set to true while publication and false while the restore flow is true.

**Comment by Arpit-Bandejiya** [comment]
This PR resolves #14109  as well.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (himshikha) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1253]: Got it.

File: server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java
Comment (himshikha) [inline server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java:80]: nit: rename to `getAsyncIndexRoutingReadAction`

File: server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java:92]: will prefer a dedicated thread pool

Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java]: Ack, will be using the same threadpool as for the remote state: `REMOTE_STATE_READ`

File: server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java]: Done.

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java
Comment (himshikha) [inline server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:308]: This should be `logger.error`

Comment (himshikha) [reply server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:308]: Also we dont need `e.toString()`, we can add exception directly here. Logger handles the conversions required by default.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java]: Removed.

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java]: Addressed.

File: server/src/test/java/org/opensearch/gateway/remote/RemoteClusterStateServiceTests.java
Comment (himshikha) [inline server/src/test/java/org/opensearch/gateway/remote/RemoteClusterStateServiceTests.java:1086]: With read part integrated, we should be able to test for indicesRouting being correctly read?

**Review by Bukhtawar — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreClusterStateRestoreIT.java
Comment (shwetathareja) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreClusterStateRestoreIT.java:330]: these are not rebased with https://github.com/opensearch-project/OpenSearch/pull/13924

Comment (shiv0408) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreClusterStateRestoreIT.java]: Updated

File: server/src/test/java/org/opensearch/gateway/remote/RemoteClusterStateServiceTests.java
Comment (shiv0408) [reply server/src/test/java/org/opensearch/gateway/remote/RemoteClusterStateServiceTests.java]: Will add ITs in a separate PR.

**Comment by shwetathareja** [comment]
PR has an approval from @Bukhtawar . Providing it another approval to unblock merge as some test fix went in since the last approval.

**Review by shwetathareja — APPROVED** [review]
Providing approval to unblock merge.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java
Comment (ashking94) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:471]: Can we please change the log level to trace or debug? This is leading to lot of logs while running a cluster on account of remote cluster state uploads.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:471]: Raised PR #14814 for this

**Pull request merged**
